### ISSUE TITLE
Rule proposal: `prefer-queue-microtask`

### ISSUE THREAD

**Issue by kevinoid** [sindresorhus/eslint-plugin-unicorn#1346]
[`queueMicrotask()`](https://nodejs.org/api/globals.html#globals_queuemicrotask_callback) serves a similar purpose to [`process.nextTick()`](https://nodejs.org/api/process.html#process_process_nexttick_callback_args) with the additional advantage that [`queueMicrotask()` is specified in the WHATWG HTML Living Standard](https://html.spec.whatwg.org/multipage/timers-and-user-prompts.html#microtask-queuing) and supported on both [modern browsers](https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/queueMicrotask#browser_compatibility) and Node.js (v11.0.0 and later).

[When to use `queueMicrotask()` vs. `process.nextTick()`](https://nodejs.org/api/process.html#process_when_to_use_queuemicrotask_vs_process_nexttick) in the Node.js API documentation states:

> For *most* userland use cases, the `queueMicrotask()` API provides a portable and reliable mechanism for deferring execution that works across multiple JavaScript platform environments and should be favored over `process.nextTick()`.  In simple scenarios, `queueMicrotask()` can be a drop-in replacement for `process.nextTick()`.

I propose adding a rule to warn when `process.nextTick` is used, and suggest `queueMicrotask()` as an alternative.

## Fail

```js
process.nextTick(() => process.exit(0));
```

```js
process.nextTick(process.exit);
```

```js
process.nextTick(process.exit, 1);
```

```js
const fun = process.nextTick;
```

```js
fun(process.nextTick);
```

## Pass

```js
queueMicrotask(() => process.exit(0));
```

```js
queueMicrotask(process.exit);
```

```js
// Note: queueMicrotask takes exactly 1 arg.
// Fail example fixed using .bind().
queueMicrotask(process.exit.bind(undefined, 1));
```

```js
// Not sure if this should be auto-fixed.
// fun may be called with multiple args.
const fun = queueMicrotask;
```

```js
// Not sure if this should be auto-fixed.
// fun may call its argument with multiple args.
fun(process.nextTick);
```

**Comment by fisker** [comment]
I never used `process.nextTick` in Node.js, but in browser, I used `setTimeout` `setImmediate` `requestAnimationFrame` `MutationObserver` `Promise#then` to do similar thing.

**Comment by sindresorhus** [comment]
https://www.freecodecamp.org/news/the-differences-between-javascripts-asynchronous-api-timers-d916e0596716/

**Comment by kevinoid** [comment]
Good points.  Lots of APIs with subtle differences.  I've also used those at different times, or for different purposes.  It's complicated.

@fisker Any reservations about a rule to prefer `queueMicrotask()` to `process.nextTick()` specifically?  Or are you considering a rule to prefer any of the others?  In either case, I'd be happy to discuss the differences and work through the details.

**Comment by fisker** [comment]
What I was saying, I think `queueMicrotask` should be preferred, but check `process.nextTick` is not going to help me at all.

**Comment by sindresorhus** [comment]
> I never used process.nextTick in Node.js, but in browser, I used setTimeout setImmediate requestAnimationFrame MutationObserver Promise#then to do similar thing.

I will try to summarize the differences as far as I know it:

- `setTimeout(..., 0)` executes at the end of the event loop run (after all tasks).
- `setImmediate()` executes after all I/O tasks.
- `requestAnimationFrame` is about rendering.
- `MutationObserver` is for DOM mutatations, but has been used as a hack to get the same behavior as `queueMicrotask`.
- `Promise#then` does almost the same as `queueMicrotask` as it adds the callback to the micro-task queue.

**Comment by sindresorhus** [comment]
We could support `Promise#then`, but it will conflict with rules enforcing `await`. And when used with `await` it solves a different problem:

```js
(async () => {
	// Cannot use `queueMicrotask` here.
	await ''; // Defer a micro-task tick.

	console.log('Hello');
})();

**Comment by sindresorhus** [comment]
`setTimeout` and `setImmediate` works differently, so changing to `queueMicrotask` might cause weird bugs or not be possible at all. We could potentially add an option to make it opt-in.

**Comment by sindresorhus** [comment]
Ideally, we would have a rule `prefer-set-immediate` which would prefer it over `setTimeout(..., 0)`, but `setImmediate` doesn't work in the browser, so it's not a good API to use.

**Comment by sindresorhus** [comment]
Seems like we'll eventually get another alternative, so we can put this rule on hold until it's a reality:

- `scheduler.yield()`: https://github.com/WICG/scheduling-apis
- https://github.com/nodejs/node/pull/40909

**Comment by sindresorhus** [comment]
> Seems like we'll eventually get another alternative, so we can put this rule on hold until it's a reality:

Doesn't look like it will happen anytime soon, so this is now accepted. Let's start with preferring `queueMicrotask` over `process.nextTick`, with opt in to also enforce for `setImmediate` and `setTimeout..., 0)`.

**Comment by woody-li** [comment]
How about add an option that only for `Promise#then`, excludes `async/await`.

**Comment by sindresorhus** [comment]
I don't plan that. async/await should be preferred. I don't see a strong enough reason to support "then".
